library/modules/Units.cpp

```diff
@@ -1063,517 +837,404 @@ void Units::setNickname(df::unit *unit, std::string nick)
-        case df::animal_training_level::SkilfullyTrained:
-        case df::animal_training_level::ExpertlyTrained:
-        case df::animal_training_level::ExceptionallyTrained:
-        case df::animal_training_level::MasterfullyTrained:
-            return training_quality_table[level-1];
-        case df::animal_training_level::Domesticated:
+        case Trained:
+            return "trained";
+        case WellTrained:
+            return "-trained-";
+        case SkilfullyTrained:
+            return "+trained+";
+        case ExpertlyTrained:
+            return "*trained*";
+        case ExceptionallyTrained:
+            return "\xF0trained\xF0"; // "≡trained≡"
+        case MasterfullyTrained:
+            return "\x0Ftrained\x0F"; // "☼trained☼"
+        case Domesticated:
             return "tame";
-        case df::animal_training_level::WildUntamed:
+        case WildUntamed:
         default:
             return "wild";
     }
 }
 
-string Units::getReadableName(df::unit* unit) {
-    string base_name = getProfessionName(unit);
-    string name = Translation::TranslateName(getVisibleName(unit), false);
-    if (name.empty()) {
-        name = base_name;
-    } else {
-        name += ", ";
-        name += base_name;
-    }
-    if (isGhost(unit)) {
-        name = "Ghostly " + name;
-    }
+string Units::getReadableName(df::historical_figure *hf) {
+    CHECK_NULL_POINTER(hf);
+    string prof_name = getProfessionName(hf, false, false, true);
 
-    for (auto unit_syndrome : unit->syndromes.active) {
-        auto syndrome = df::syndrome::find(unit_syndrome->type);
-        if (!syndrome)
-            continue;
-        for (auto effect : syndrome->ce) {
-            auto cie = strict_virtual_cast<df::creature_interaction_effect_display_namest>(effect);
-            if (!cie)
-                continue;
-            name += " ";
-            name += cie->name;
-            break;
-        }
+    if (hf->info && hf->info->curse) {
+        auto &curse = *hf->info->curse;
+        if (curse.original_race >= 0 && !curse.undead_name.empty()) // Zombie
+            prof_name = curse.undead_name; // Replaces entire profession
+        else if (curse.use_display_name && !curse.name.empty()) // Necromancer, etc.
+            prof_name += " " + curse.name;
+        else if (hf->flags.is_set(histfig_flags::ghost))
+            prof_name = "Ghostly " + prof_name;
     }
 
-    if (unit->enemy.undead) {
+    string name = Translation::TranslateName(getVisibleName(hf), false);
+    return name.empty() ? prof_name : name + ", " + prof_name;
+}
+
+string Units::getReadableName(df::unit *unit) {
+    CHECK_NULL_POINTER(unit);
+    string prof_name = getProfessionName(unit, false, false, true);
+
+    if (unit->curse.name_visible) // Necromancer, etc.
+        prof_name += " " + unit->curse.name;
+    else if (isGhost(unit)) // TODO: Should be "Ghost" instead of "Ghostly Peasant"
+        prof_name = "Ghostly " + prof_name;
```

Comment:
This is a behavior change -- are you sure that curse and ghost are mutually exclusive? Can you not have names like "Ghostly human necromancer"?

Reply:
![ghostly](https://github.com/user-attachments/assets/ea1cb733-5690-4b13-9f8b-57f2bde48f5a)